atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java

```diff
@@ -77,11 +97,59 @@ public ListenableFuture<Long> get(Long startTimestamp) {
      * */
     @Override
     public ListenableFuture<Map<Long, Long>> get(Iterable<Long> keys) {
-        Map<Long, ListenableFuture<Long>> futures = KeyedStream.of(
-                        StreamSupport.stream(keys.spliterator(), false).sorted(Comparator.reverseOrder()))
-                .map(this::getInternal)
-                .collectToMap();
-        return AtlasFutures.allAsMap(futures, MoreExecutors.directExecutor());
+        if (IS_VALIDATION_MODE) {
+            // No verification for in-progress transactions right now.
+            return Futures.transform(
+                    delegate.get(keys),
+                    statuses -> KeyedStream.stream(statuses)
+                            .map(this::verifyAndGetCommitTs)
+                            .flatMap(Optional::stream)
+                            .collectToMap(),
+                    MoreExecutors.directExecutor());
+        } else {
+            Map<Long, ListenableFuture<Long>> futures = KeyedStream.of(
+                            StreamSupport.stream(keys.spliterator(), false).sorted(Comparator.reverseOrder()))
+                    .map(this::getInternal)
+                    .collectToMap();
+            return AtlasFutures.allAsMap(futures, MoreExecutors.directExecutor());
+        }
+    }
+
+    private Optional<Long> verifyAndGetCommitTs(long startTimestamp, TransactionStatus transactionStatus) {
+        Optional<Long> maybeGetCommitTs = Optional.ofNullable(TransactionStatusUtils.getCommitTsFromStatus(
+                startTimestamp, transactionStatus, knownAbandonedTransactions::isKnownAbandoned));
+
+        maybeGetCommitTs.ifPresent(commitTs -> {
+            boolean isConcluded = knownConcludedTransactions.isKnownConcluded(
+                    startTimestamp, KnownConcludedTransactions.Consistency.REMOTE_READ);
+
+            if (!isConcluded) {
```

Comment:
We can also check even if not concluded that we don't get abandoned && committed

Reply:
The thing is we cannot ask abandoned cache for status of a non concluded timestamp. So we will have to talk to the store and that would carry perf penalty.